src/java/org/apache/cassandra/journal/SyncedOffsets.java

```diff
@@ -0,0 +1,261 @@
+ * limitations under the License.
+ */
+package org.apache.cassandra.journal;
+
+import java.io.IOException;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.file.StandardOpenOption;
+import java.util.zip.CRC32;
+
+import com.google.common.primitives.Ints;
+
+import org.apache.cassandra.io.util.File;
+import org.apache.cassandra.io.util.FileUtils;
+import org.apache.cassandra.io.util.RandomAccessReader;
+import org.apache.cassandra.utils.Closeable;
+import org.apache.cassandra.utils.Crc;
+import org.apache.cassandra.utils.SyncUtil;
+
+import static org.apache.cassandra.utils.FBUtilities.updateChecksumInt;
+
+/**
+ * Keeps track of fsynced limits of a data file. Enables us to treat invalid
+ * records that are known to have been fsynced to disk differently from those
+ * that aren't.
+ * <p/>
+ * On disk representation is a sequence of 2-int tuples of:
+ *      (synced offset, CRC32(tuple position in synced offsets file, synced offset))
+ */
+interface SyncedOffsets extends Closeable
+{
+    /**
+     * @return furthest known synced offset
+     */
+    int syncedOffset();
+
+    /**
+     * Record an offset as synced to disk.
+     *
+     * @param offset the offset into datafile, up to which contents have been fsynced (exclusive)
+     */
+    void mark(int offset);
+
+    @Override
+    default void close()
+    {
+    }
+
+    /**
+     * @return a mutable MMAP-backed synced offset tracker for a new {@link ActiveSegment}
+     */
+    static Active active(Descriptor descriptor, boolean syncOnMark)
+    {
+        return new Active(descriptor, syncOnMark);
+    }
+
+    /**
+     * Load an existing log of synced offsets from disk into an immutable instance.
+     */
+    static Static load(Descriptor descriptor)
+    {
+        return Static.load(descriptor);
+    }
+
+    /**
+     * @return a placeholder instance in case this component is missing
+     */
+    static Absent absent()
+    {
+        return Absent.INSTANCE;
+    }
+
+    /**
+     * Single-threaded, MMAP-backed list of synced offsets.
+     */
+    final class Active implements SyncedOffsets
+    {
+        private static final int INITIAL_CAPACITY = 4 << 10;
+
+        private final Descriptor descriptor;
+        private final boolean syncOnMark;
+
+        private final FileChannel channel;
+        private MappedByteBuffer buffer;
+
+        private volatile int syncedOffset;
+
+        private Active(Descriptor descriptor, boolean syncOnMark)
+        {
+            this.descriptor = descriptor;
+            this.syncOnMark = syncOnMark;
+
+            File file = descriptor.fileFor(Component.SYNCED_OFFSETS);
+            if (file.exists())
+                throw new IllegalArgumentException("Synced offsets file " + file + " already exists");
+
+            try
+            {
+                channel = FileChannel.open(file.toPath(), StandardOpenOption.WRITE, StandardOpenOption.READ, StandardOpenOption.CREATE);
```

Comment:
This causes issues with simulator 

```
INFO  [AsyncAppender-Worker-ASYNC] 2023-04-12 15:21:40,572 SubstituteLogger.java:169 - ERROR [AccordJournal-allocator:1] 2023-04-12 15:21:40,572 Journal.java:615 - Failed allocating journal segments. Journal AccordJournal failure policy is STOP; terminating thread.
java.lang.UnsupportedOperationException: null
	at com.google.common.jimfs.JimfsFileChannel.map(JimfsFileChannel.java:598)
	at org.apache.cassandra.journal.SyncedOffsets$Active.map(SyncedOffsets.java:164)
	at org.apache.cassandra.journal.SyncedOffsets$Active.<init>(SyncedOffsets.java:120)
	at org.apache.cassandra.journal.SyncedOffsets$Active.<init>(SyncedOffsets.java:91)
	at org.apache.cassandra.journal.SyncedOffsets.active(SyncedOffsets.java:69)
	at org.apache.cassandra.journal.ActiveSegment.create(ActiveSegment.java:83)
	at org.apache.cassandra.journal.Journal.createSegment(Journal.java:450)
	at org.apache.cassandra.journal.Journal.access$200(Journal.java:75)
	at org.apache.cassandra.journal.Journal$AllocateRunnable.runNormal(Journal.java:392)
	at org.apache.cassandra.journal.Journal$AllocateRunnable.run(Journal.java:374)
```

Reply:
Resolved now. Am aware of the two other cases.